File: app/src/main/java/com/example/ruslanyussupov/popularmovies/list/MovieGridFragment.java
package com.example.ruslanyussupov.popularmovies.list;


import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.ruslanyussupov.popularmovies.ItemDecoration;
import com.example.ruslanyussupov.popularmovies.R;
import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter;
import com.example.ruslanyussupov.popularmovies.databinding.FragmentMovieGridBinding;

import java.util.ArrayList;
import java.util.Collections;

import timber.log.Timber;

import static com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter.*;


public class MovieGridFragment extends Fragment {

    public static final String EXTRA_MOVIE = "movie";

    private static final int MOVIE_GRID_COLUMNS = 2;

    private MovieAdapter mMovieAdapter;
    private OnMovieClickListener mMovieClickListener;
    private FragmentMovieGridBinding mBinding;
    private MainViewModel mViewModel;

    public MovieGridFragment() {}

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mMovieClickListener = (OnMovieClickListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement OnMovieClickListener");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        Timber.d("onCreateView");
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_movie_grid, container, false);
        return mBinding.getRoot();

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        Timber.d("onActivityCreated");

        mMovieAdapter = new MovieAdapter(new ArrayList<>(), mMovieClickListener);
        mBinding.rvMovies.setAdapter(mMovieAdapter);
        mBinding.rvMovies.setLayoutManager(new GridLayoutManager(getActivity(), MOVIE_GRID_COLUMNS));
        int offset = getResources().getDimensionPixelOffset(R.dimen.movie_item_offset);
        mBinding.rvMovies.addItemDecoration(new ItemDecoration(offset, offset, offset, offset));

        mViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);

        mViewModel.getResultLiveData().observe(this, result -> {

            mBinding.rvMovies.scrollToPosition(0);

            switch (result.state) {
                case LOADING:
                    Timber.d("Movies loading...");
                    mBinding.loadingPb.setVisibility(View.VISIBLE);
                    break;
                case SUCCESS:
                    mBinding.loadingPb.setVisibility(View.GONE);
                    if (result.data == null || result.data.isEmpty()) {
                        Timber.d("Result is empty.");
                        showSnackBar("No movies!");
                        mMovieAdapter.updateData(Collections.emptyList());
                    } else {
                        Timber.d("Movies loaded successfully: %s", result.data);
                        mMovieAdapter.updateData(result.data);
                    }
                    break;
                case ERROR:
                    mBinding.loadingPb.setVisibility(View.GONE);
                    if (mViewModel.getUtils().hasNetworkConnection()) {
                        Timber.e("Error while loading movies: %s", result.error);
                        showSnackBar(result.error);
                    } else {
                        Timber.w("No network connection.");
                        showSnackBar("No connection!");
                    }
                    break;
            }

        });

    }

    @Override
    public void onDestroy() {
        Timber.d("onDestroy");
        super.onDestroy();
    }

    private void showSnackBar(String message) {
        Snackbar.make(mBinding.rvMovies, message, Snackbar.LENGTH_LONG)
                .setAction("Retry", listener -> mViewModel.retry())
                .show();
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/list/MovieGridFragment.kt
package com.example.ruslanyussupov.popularmovies.list


import androidx.lifecycle.ViewModelProviders
import android.content.Context
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import com.google.android.material.snackbar.Snackbar
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer

import com.example.ruslanyussupov.popularmovies.ItemDecoration
import com.example.ruslanyussupov.popularmovies.R
import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter
import com.example.ruslanyussupov.popularmovies.databinding.FragmentMovieGridBinding

import timber.log.Timber

import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter.*
import com.example.ruslanyussupov.popularmovies.data.model.Movie
import com.example.ruslanyussupov.popularmovies.Result


class MovieGridFragment : Fragment() {

    private lateinit var movieAdapter: MovieAdapter
    private lateinit var movieClickListener: OnMovieClickListener
    private lateinit var binding: FragmentMovieGridBinding
    private lateinit var viewModel: MainViewModel

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnMovieClickListener) {
            movieClickListener = context
        } else {
            throw ClassCastException("$context must implement OnMovieClickListener")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        Timber.d("onCreateView")
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_movie_grid, container, false)
        return binding.root

    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        Timber.d("onActivityCreated")

        movieAdapter = MovieAdapter(emptyList(), movieClickListener::onMovieClick)
        binding.rvMovies.adapter = movieAdapter
        binding.rvMovies.layoutManager = GridLayoutManager(activity, MOVIE_GRID_COLUMNS)
        val offset = resources.getDimensionPixelOffset(R.dimen.movie_item_offset)
        binding.rvMovies.addItemDecoration(ItemDecoration(offset, offset, offset, offset))

        viewModel = ViewModelProviders.of(activity!!).get(MainViewModel::class.java)

        viewModel.getResultLiveData().observe(this, Observer<Result<List<Movie>>> { result ->

            binding.rvMovies.scrollToPosition(0)

            when (result.state) {
                Result.State.LOADING -> {
                    Timber.d("Movies loading...")
                    binding.loadingPb.visibility = View.VISIBLE
                }
                Result.State.SUCCESS -> {
                    binding.loadingPb.visibility = View.GONE
                    if (result.data == null || result.data.isEmpty()) {
                        Timber.d("Result is empty.")
                        showSnackBar("No movies!")
                        movieAdapter.updateData(emptyList())
                    } else {
                        Timber.d("Movies loaded successfully: ${result.data}")
                        movieAdapter.updateData(result.data)
                    }
                }
                Result.State.ERROR -> {
                    binding.loadingPb.visibility = View.GONE
                    if (viewModel.utils.hasNetworkConnection()) {
                        Timber.e("Error while loading movies: ${result.error}")
                        showSnackBar(result.error)
                    } else {
                        Timber.w("No network connection.")
                        showSnackBar("No connection!")
                    }
                }
            }

        })

    }

    override fun onDestroy() {
        Timber.d("onDestroy")
        super.onDestroy()
    }

    private fun showSnackBar(message: String) {
        Snackbar.make(binding.rvMovies, message, Snackbar.LENGTH_LONG)
                .setAction("Retry") { viewModel.retry() }
                .show()
    }

    companion object {
        const val EXTRA_MOVIE = "movie"
        private const val MOVIE_GRID_COLUMNS = 2
    }

}
